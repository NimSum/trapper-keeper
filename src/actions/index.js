export const addNotes = notes => ({
  type: 'ADD_NOTES',
  notes
});

export const addNote = note => ({
  type: 'ADD_NOTE',
  note
})

export const updateNote = note => ({
  type: 'UPDATE_NOTE',
  note
})
