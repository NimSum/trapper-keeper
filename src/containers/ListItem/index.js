import React, { Component } from 'react'
import { connect } from 'react-redux';

export class ListItem extends Component {
  constructor() {
    super();
    this.state = {
      editable: false,
      body: ''
    }
  }

  componentDidMount() {
    this.setState({ body: this.props.item.body })
  }

  editItem = () => {
    this.setState({ editable: true })
  }

  handleChange = ({ target }) => {
    this.setState({ body: target.value })
  }

  handleSubmit = e => {
    e.preventDefault();
    this.setState({ editable: false });
    const { id, completed } = this.props.item;
    const { body } = this.state;
    this.props.updateListItems({
      id,
      completed,
      body
    })
  }

  checkItem = () => {
    const { id, completed, body } = this.props.item;
    this.props.updateListItems({
      id,
      completed: !completed,
      body
    })
  }
  
  deleteItem =  () => {
    this.props.updateListItems({...this.props.item}, true)
  }
  
  render() {
    const { id, completed } = this.props.item;
    console.log(this.props.editing)
    const form = (
      <form onSubmit={ this.handleSubmit }>
        <input
          className= 'list-item-input' 
          type="text" 
          value={this.state.body}
          onChange={ this.handleChange }
          onBlur={ this.handleSubmit }
          autoFocus />
      </form>)
    return (
      <li 
        key={ id }>
        { this.state.editable 
            ? form
            : (
              <div className="item-container">
                <button 
                  onClick={ this.checkItem }>
                  done
                </button>
                <p 
                  onClick={ this.props.editing && this.editItem }
                  className={ completed 
                    ? 'completed-item list-item' 
                    : 'list-item'}>    
                    {this.props.item.body || this.state.body }
                </p>
                <button onClick={ this.deleteItem }>x</button>
              </div>
              )
        }
      </li>
    )
  }
}


const mapDispatchToProps = dispatch => ({
})

export default connect(null, mapDispatchToProps)(ListItem);