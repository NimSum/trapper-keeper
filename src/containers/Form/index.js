import React, { Component } from 'react'
import ListItem from '../ListItem';
import { addNewNote } from '../../thunks/addNewNote';
import { connect } from 'react-redux';
import { updateNote } from '../../actions/index';
import { PropTypes } from 'prop-types';
import  uuidv4 from 'uuid/v4';
import { putNote } from '../../utils/apiFetches/putNote';

export class Form extends Component {
  constructor() {
    super();
    this.state = {
      title: '',
      listItemText: '',
      listItems: [],
      editing: false,
      id: ''
    };
  }

  componentDidMount() {
    if(this.props.foundNote) {
      const { id, title, listItems } = this.props.foundNote;
      this.setState({ id, title, listItems, editing: true })
    }
  }

  static getDerivedStateFromProps(props, state) {
    if (!props.foundNote) return null;
    if (props.foundNote.id !== state.id) {
      const { id, title, listItems } = props.foundNote;
      return { id, title, listItems, editing: true }
    } else return null;
  }

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value })
  }
  
  handleSubmit = e => {
    e.preventDefault();
    const { listItemText } = this.state;
    const newItem = { id: uuidv4(), body: listItemText, completed: false }
      listItemText.length && this.setState({ 
      listItems: [...this.state.listItems, newItem],
      listItemText: ''
    })
  }

  editNote = async () => {
    const { id, listItems, title} = this.state;
    try {
      await putNote({ id, listItems, title });
      this.props.updateExistingNote({ id, listItems, title });
    } catch(error) {
      console.log(error)
    }
    this.setState({ editing: false })
  }

  addNote = () => {
    const { title, listItems } = this.state;
    const newNote = {
      title,
      listItems
    }
    this.props.addNewNote(newNote);
  }

  updateListItems = (newItem, remove) => {
    let updatedListItems;
    if (remove) {
      updatedListItems = this.state.listItems.filter(listItem => {
        return listItem.id !== newItem.id
      });
    } else {
      updatedListItems = [...this.state.listItems].map(listItem => {
        if (listItem.id === newItem.id) {
          return newItem;
        } else return listItem;
      })
    } 
    this.setState({ listItems: updatedListItems });
  }

  render() {
    return (
      <div className='form-container'>
        <div className='title-form'>
          <input 
            placeholder='Title' 
            type='text'
            name='title'
            value={ this.state.title }
            onChange={ this.handleChange } 
            className='title'
          /> 
        </div>
        <div>
          {this.state.listItems.map(item => {
            return (
            <ListItem 
              item={ item }
              editing={ this.state.editing }
              updateListItems={ this.updateListItems } />
          )})}
        </div>
        <form onSubmit={ this.handleSubmit }>
          <input 
            type="text" 
            name="listItemText"
            value={this.state.listItemText}
            onChange={ this.handleChange }
            onBlur={ this.handleSubmit }
            autoFocus />
        </form>
        <button onClick={ () => this.state.editing ? this.editNote() : this.addNote() }>Save</button>
      </div>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  addNewNote: (note) => dispatch(addNewNote(note)),
  updateExistingNote: (note) => dispatch(updateNote(note))
})

export default connect(null, mapDispatchToProps)(Form);
