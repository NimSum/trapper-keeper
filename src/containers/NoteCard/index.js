import React, { Component } from 'react'
import ListItem from '../../containers/ListItem';
import { Link } from 'react-router-dom'
import { updateNote } from '../../actions/index';
import { connect } from 'react-redux';
import {deleteNote} from '../../actions';
import {deleteNoteFetch} from '../../utils/apiFetches/deleteNote';
export class NoteCard extends Component {
  constructor() {
    super();
    this.state = {
    }
  }

  componentDidMount() {
    this.setState({ ...this.props.note })
  }

  updateListItems = (newItem, remove) => {
    // make updated note with list item
    let updateListItems;
    if (remove) {
      updateListItems = [...this.state.listItems].filter(listItem => listItem.id !== newItem.id);
    } else {
      updateListItems = [...this.state.listItems].map(listItem => {
        if (listItem.id === newItem.id) {
          return newItem;
        } else return listItem;
      })
    } 

    this.setState({ listItems: updateListItems }, () => 
      this.props.updateNote({...this.state})
    );
  }

  deleteNote = async() => {
    const {id} = this.props.note
    console.log(id, 'Testing delete')
    try{
      deleteNoteFetch(id);
      this.props.removeNote(id)
    }catch(error){
      console.log('deleteNote', error)
    }
  }

  render() {
    const listItems = this.props.note.listItems.map(item => (
      < ListItem updateListItems={ this.updateListItems } item={ item } />
      // <p className={ item.completed ? 'completed-item list-item' : 'list-item'}>
      //   { item.body }
      // </p>
    ))

    return (
      <Link exact to={`/notes/${this.props.note.id}`} style={{ textDecoration: 'none'}}>
        <article className='note-card'>
          <h3>{this.props.note.title}</h3>
          <ul className='note-items'>
            { listItems }
          </ul>
          <button className="delete-card" onClick={this.deleteNote}>X</button>
        </article>
      </Link>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  updateNote: note => dispatch(updateNote(note)),
  removeNote: id=> dispatch(deleteNote(id))
})

export default connect(null, mapDispatchToProps)(NoteCard);
