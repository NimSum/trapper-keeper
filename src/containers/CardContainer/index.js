import React from 'react';
import { connect } from 'react-redux';
import NoteCard from '../../containers/NoteCard';
import PropTypes from 'prop-types';


export const CardContainer = ({ notes }) => (
  <section className="notes-container">
    { notes.map(note => {
      return < NoteCard 
        note={ note } 
        key={note.id} />
    })}
    
  </section>
)

export const mapStateToProps = (state) => ({
  notes: state.notes
})

CardContainer.propTypes = {
  dispatch: PropTypes.func, 
  notes: PropTypes.array
}

export default connect(mapStateToProps)(CardContainer);