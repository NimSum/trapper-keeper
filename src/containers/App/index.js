import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addNotes } from '../../actions/';
import CardContainer from '../../components/CardContainer';
import { Route } from 'react-router-dom'
export class App extends Component {

  async componentDidMount() {
    const response = await fetch('http://localhost:3000/api/v1/notes');
    const notes = await response.json();
    this.props.setNotes(notes.notes)
  } 

  render() {
    return (
      <div>
        <h1>Trapper Keeper</h1>
        < CardContainer />
        <Route path='/new-note' component={Form} />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({

})

const mapDispatchToProps = (dispatch) => ({
  setNotes: (notes) => dispatch(addNotes(notes))
})

export default connect(mapStateToProps, mapDispatchToProps)(App);
