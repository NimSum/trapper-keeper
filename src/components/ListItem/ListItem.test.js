import React from "react";
import ReactDOM from "react-dom";
import { ListItem } from "./index.js";
import { shallow, mount } from "enzyme";
import { mapDispatchToProps } from "../../containers/App";
import { addNotes } from "../../actions/index";

describe("ListItem", () => {
  let wrapper;
  let mockNotes;
  let mockItem;
  let mockEditing = false;
  let mockUpdateListItems = jest.fn();
  let mockDefaultState = {
    editable: false,
    body: ""
  };

  beforeEach(() => {
    mockItem = {
      id: 4,
      body: "Hello",
      completed: false
    };
    wrapper = shallow(
      <ListItem
        item={mockItem}
        editing={mockEditing}
        updateListItems={mockUpdateListItems}
      />,
      { disableLifecycleMethods: true }
    );
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot when editable state is true", () => {
    wrapper.setState({ editable: true });
    expect(wrapper).toMatchSnapshot();
  })

  it("should have a default state", () => {
    expect(wrapper.state()).toEqual(mockDefaultState);
  });

  it("should set the editable property to true", () => {
    wrapper.editItem = jest.fn();
    wrapper.instance().editItem();
    expect(wrapper.state()).toEqual({ body: "", editable: true });
  });

  it("should update state of body on handleChange", () => {
    const mockEvent = { target: { value: "Clean Room" } };
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state("body")).toBe("Clean Room");
  });

  it("should set state of editable to false on handleSubmit", () => {
    const mockEvent = { target: { value: false } };
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state("editable")).toBe(false);
  });

  it("should call updateListItems on handleSubmit", () => {
    const mockEvent = { preventDefault: jest.fn() };
    wrapper.instance().handleSubmit(mockEvent);
    expect(mockUpdateListItems).toHaveBeenCalled();
  });

  it("should call updateListItems on checkItem", () => {
    wrapper.instance().checkItem();
    expect(mockUpdateListItems).toHaveBeenCalled();
  });

  it("should call updateListItems on deleteItem", () => {
    wrapper.instance().deleteItem();
    expect(mockUpdateListItems).toHaveBeenCalled();
  });

  it("should match snapshot when isEditing is true", () => {
    wrapper = shallow(
      <ListItem
        item={mockItem}
        editing={!mockEditing}
        updateListItems={mockUpdateListItems}
      />,
      { disableLifecycleMethods: true }
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot when completed is set to true", () => {
    mockItem = {
      id: 4,
      body: "Hello",
      completed: true
    };
    wrapper = shallow(
      <ListItem
        item={mockItem}
        editing={!mockEditing}
        updateListItems={mockUpdateListItems}
      />,
      { disableLifecycleMethods: true }
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should invoke editing when list item is clicked', () => {
    wrapper.setProps({ editing: true })
    wrapper.find('p.list-item').simulate('click');
    expect(wrapper.state().editable).toBe(true);
  })


});
