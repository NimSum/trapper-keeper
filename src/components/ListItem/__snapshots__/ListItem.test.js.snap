// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should match snapshot when completed is set to true 1`] = `
<li
  className="list-item"
  key="4"
>
  <div
    className="item-container"
  >
    <i
      className="far fa-check-square"
      onClick={[Function]}
    />
    <p
      className="completed-item list-item"
      onClick={[Function]}
    >
      Hello
    </p>
    <i
      className="fas fa-times"
      onClick={[Function]}
    />
  </div>
</li>
`;

exports[`ListItem should match snapshot when editable state is true 1`] = `
<li
  className="list-item"
  key="4"
>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="list-item-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
  </form>
</li>
`;

exports[`ListItem should match snapshot when isEditing is true 1`] = `
<li
  className="list-item"
  key="4"
>
  <div
    className="item-container"
  >
    <i
      className="far fa-square"
      onClick={[Function]}
    />
    <p
      className="list-item"
      onClick={[Function]}
    >
      Hello
    </p>
    <i
      className="fas fa-times"
      onClick={[Function]}
    />
  </div>
</li>
`;

exports[`ListItem should match the snapshot 1`] = `
<li
  className="list-item"
  key="4"
>
  <div
    className="item-container"
  >
    <i
      className="far fa-square"
      onClick={[Function]}
    />
    <p
      className="list-item"
      onClick={false}
    >
      Hello
    </p>
    <i
      className="fas fa-times"
      onClick={[Function]}
    />
  </div>
</li>
`;
